src/cmd/services/m3coordinator/downsample/tags.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package downsample
+
+import (
+	"sort"
+
+	"github.com/m3db/m3x/ident"
+)
+
+type tags struct {
+	names    []string
+	values   []string
+	idx      int
+	nameBuf  []byte
+	valueBuf []byte
+}
+
+var _ ident.TagIterator = &tags{}
+var _ sort.Interface = &tags{}
+
+func newTags() *tags {
+	return &tags{
+		names:  make([]string, 0, initAllocTagsSliceCapacity),
+		values: make([]string, 0, initAllocTagsSliceCapacity),
+		idx:    -1,
+	}
+}
+
+func (t *tags) Len() int {
+	return len(t.names)
+}
+
+func (t *tags) Swap(i, j int) {
+	t.names[i], t.names[j] = t.names[j], t.names[i]
+	t.values[i], t.values[j] = t.values[j], t.values[i]
+}
+
+func (t *tags) Less(i, j int) bool {
+	return t.names[i] < t.names[j]
+}
+
+func (t *tags) Next() bool {
+	hasNext := t.idx+1 < len(t.names)
```

Comment:
seems like other functions in this package are appending to t.names. In that case, Next() can return false first but then return true which makes it a bit confusing.

Reply:
You're meant to construct it first, then use it as an iterator.  I don't want to create intermediate structures or else they can't be pooled easily.